### PR TITLE
Typed exceptions

### PR DESCRIPTION
**Pull request by ehildenb**
Simplest steps towards typed exceptions, makes refactoring the control flow simpler.

### CODE REVIEW

**Comment by ehildenb** [comment]
VMTests disabled because: https://github.com/ethereum/tests/pull/444

**Comment by ehildenb** [comment]
@dwightguth should be ready for review. The failing proofs are because the proof submodule has not been updated, doing that now but they should go through fine once that's done.

**Comment by dwightguth** [comment]
Jenkins: test mantis

**Comment by ehildenb** [comment]
Jenkins: test this please

**Comment by ehildenb** [comment]
Jenkins: test mantis

**Comment by ehildenb** [comment]
Jenkins: test this please

File: network.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+The following codes all indicate that the VM ended execution with an exception, but give details about how.
+
+-   `EVMC_FAILURE` is a catch-all for generic execution failure.
+-   `EVMC_INVALID_INSTRUCTION` indicates reaching the designated `INVALID` opcode.
+-   `EVMC_UNDEFINED_INSTRUCTION` indicates that an undefined opcode has been reached.
+-   `EVMC_OUT_OF_GAS` indicates that execution exhausted the gas supply.
+-   `EVMC_BAD_JUMP_DESTINATION` indicates a `JUMP*` to a non-`JUMPDEST` location.
+-   `EVMC_STACK_OVERFLOW` indicates pushing more than 1024 elements onto the wordstack.
+-   `EVMC_STACK_UNDERFLOW` indicates popping elements off an empty wordstack.
+-   `EVMC_CALL_DEPTH_EXCEDED` indicates that we have executed too deeply a nested sequence of `CALL*` or `CREATE` opcodes.
```
Comment (dwightguth) [inline network.md]: pretty sure exceeded has two es.

**Comment by ehildenb** [comment]
Jenkins: test this please

**Pull request merged**